libraries/chain/include/eosio/chain/apply_context.hpp

```diff
@@ -185,7 +187,196 @@ class apply_context {
 
       void checktime() const;
 
+      void update_db_usage( const account_name& payer, int64_t delta ) {
+         require_write_lock( payer );
+         if( (delta > 0) && payer != account_name(receiver) ) {
+            require_authorization( payer );
+         }
+      }
+
+
+      int db_store_i64( uint64_t scope, uint64_t table, const account_name& payer, uint64_t id, const char* buffer, size_t buffer_size ) {
+         require_write_lock( scope );
+         const auto& tab = find_or_create_table( scope, receiver, table );
+         auto tableid = tab.id;
+
+         const auto& obj = mutable_db.create<key_value_object>( [&]( auto& o ) {
+            o.t_id        = tableid;
+            o.primary_key = id;
+            o.value.resize( buffer_size );
+            o.payer       = payer;
+            memcpy( o.value.data(), buffer, buffer_size );
+         });
+
+         mutable_db.modify( tab, [&]( auto& t ) {
+           ++t.count;
+         });
+
+         update_db_usage( payer, buffer_size + 200 );
+
+         keyval_cache.cache_table( tab );
+         return keyval_cache.add( obj );
+      }
+
+      void db_update_i64( int iterator, account_name payer, const char* buffer, size_t buffer_size ) {
+         const key_value_object& obj = keyval_cache.get( iterator );
+
+         require_write_lock( keyval_cache.get_table( obj.t_id ).scope );
+
+         int64_t old_size = obj.value.size();
+
+         if( payer == account_name() ) payer = obj.payer;
+
+         if( account_name(obj.payer) == payer ) {
+            update_db_usage( obj.payer, buffer_size + 200 - old_size );
+         } else  {
+            update_db_usage( obj.payer,  -(old_size+200) );
+            update_db_usage( payer,  (buffer_size+200) );
+         }
+
+         mutable_db.modify( obj, [&]( auto& o ) {
+           o.value.resize( buffer_size );
+           memcpy( o.value.data(), buffer, buffer_size );
+           o.payer = payer;
+         });
+      }
+
+      void db_remove_i64( int iterator ) {
+         const key_value_object& obj = keyval_cache.get( iterator );
```

Comment:
Not sure on whether we are worrying about locking yet, but we are using a reference to an object that could be removed by another thread calling db_remove_i64, right?